src/monitoring_service/database.py

```diff
@@ -1,51 +1,229 @@
+    """ DB shared by MS and request collector """
+
+    def __init__(self, filename: str):
+        self.conn = sqlite3.connect(
+            filename,
+            detect_types=sqlite3.PARSE_DECLTYPES,
+        )
+        self.conn.row_factory = sqlite3.Row
+        self.conn.execute("PRAGMA foreign_keys = ON")
 
     def upsert_monitor_request(self, request: MonitorRequest) -> None:
-        try:
-            index = self.monitor_requests.index(request)
-            self.monitor_requests[index] = request
-        except ValueError:
-            self.monitor_requests.append(request)
+        values = [
+            hex(request.channel_identifier),
+            request.token_network_address,
+            request.balance_hash,
+            hex(request.nonce),
+            request.additional_hash,
+            request.closing_signature,
+            request.non_closing_signature,
+            hex(request.reward_amount),
+            request.reward_proof_signature,
+            request.non_closing_signer,
+        ]
+        upsert_sql = "INSERT OR REPLACE INTO monitor_request VALUES ({})".format(
+            ', '.join('?' * len(values)),
+        )
+        self.conn.execute(upsert_sql, values)
 
     def get_monitor_request(
             self,
             token_network_address: str,
             channel_id: int,
+            non_closing_signer: str,
     ) -> Optional[MonitorRequest]:
-        for mr in self.monitor_requests:
-            if (
-                mr.token_network_address == token_network_address and
-                mr.channel_identifier == channel_id
-            ):
-                return mr
+        row = self.conn.execute(
+            """
+                SELECT * FROM monitor_request
+                WHERE channel_identifier = ?
+                  AND token_network_address = ?
+                  AND non_closing_signer = ?
+            """,
+            [hex(channel_id), token_network_address, non_closing_signer],
+        ).fetchone()
+        if row is None:
+            return None
+
+        kwargs = {
+            key: val for key, val in zip(row.keys(), row)
+            if key in [f.name for f in dataclasses.fields(MonitorRequest)]
+        }
+        mr = MonitorRequest(chain_id=1, **kwargs)
+        return mr
+
+    def monitor_request_count(self) -> int:
+        return self.conn.execute("SELECT count(*) FROM monitor_request").fetchone()[0]
+
+    def upsert_channel(self, channel: Channel) -> None:
+        values = [
+            channel.token_network_address,
+            hex(channel.identifier),
+            channel.participant1,
+            channel.participant2,
+            hex(channel.settle_timeout),
+            channel.state,
+            hex(channel.closing_block) if channel.closing_block else None,
+            channel.closing_participant,
+            channel.closing_tx_hash,
+            channel.claim_tx_hash,
+            dataclasses.astuple(channel.update_status) if channel.update_status else None,
```

Comment:
We could also flatten these into the channel table, not sure what's easiest here.

Reply:
I thought about flattening, too. But I had too many half-finished changes at that point, so I didn't want to introduce changes in even more places. I can do so, now.

Reply:
Doesn't need to be for this PR, and just if you feel it's worth the work. Otherwise we can see how this evolves (maybe more stuff gets into the `update_status`)